{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Profile plot description:\n",
    "The figure was generated using the profile plot utility of deeptools. Both sample show a peak in ChIP signal and an immediate drop off at the TSS and low signal all the way until the TES as expected. This shows open chromatin at the TSS as expected for conditions with high TF activity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Methods:\n",
    "Obtained reads were first put through QC analysis with FastQC. After inspecting QC metrics, reads were then trimmed using Trimmomatic before alignment with Bowtie2 to an index built from the human reference genome (GRCh38), also using Bowtie2. Reads were then sorted and indexed with samtools before samtools flagstat was used to generate metrics for alignment. At this point QC reports and log files were compiled using MultiQC. BigWig files were then generated using deeptools bamCoverage utility, with coverage summarized using deeptools bamCoverage and corrPlot utilities. Peakcalling with MACS3 was then performed on the sorted BAM files. The bedtools intersect utility was then used to generate a set of reproducible peaks by identifying peaks from both replicates that overlapped by 50% reciprocally befor again using bedtools intersect to filter out peaks overlapping blacklisted regions. Peak annotation was done using HOMER with default parameters. Bedtools computeMatrix and plotProfile utilities were then used to generate a profile plot from BigWig files. Lastly, the HOMER findMotifsGenome.pl script was used to find motifs in the filtered peaks. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "QC Report Summary:\n",
    "Firstly, looking at the FastQC reports, GC content appears to follow a roughly normal distribution but differs slightly between INPUT and IP samples, as well as having what looks to be a small second peak in all distributions. Given that we expect different sequences between IP and INPUT, this isn't a red flag. However, FastQC does consider the distributions for the INNPUT samples to show an issue, which may need to be investigated further. Quality scores also look fine, with high quality across all samples. Duplication is higher in IP samples but again that is to be expected. One concern is the much lower number of sequences in the rep2 INPUT, with 10M instead of the ~30M in other samples. Flagstat results don't seem to show red flags either, with most reads from all samples passing QC checks and mapping. Trimmomatic also shows not many sequences were dropped, which is a good sign. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import pybedtools\n",
    "import re\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>signalValue</th>\n",
       "      <th>pValue</th>\n",
       "      <th>qValue</th>\n",
       "      <th>peak</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>195221</td>\n",
       "      <td>195358</td>\n",
       "      <td>rep1_peak_216</td>\n",
       "      <td>28</td>\n",
       "      <td>.</td>\n",
       "      <td>2.69423</td>\n",
       "      <td>2.80547</td>\n",
       "      <td>0.720430</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>155811</td>\n",
       "      <td>155924</td>\n",
       "      <td>rep1_peak_280</td>\n",
       "      <td>14</td>\n",
       "      <td>.</td>\n",
       "      <td>1.55919</td>\n",
       "      <td>1.47921</td>\n",
       "      <td>0.251322</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL000208.1</td>\n",
       "      <td>47204</td>\n",
       "      <td>47410</td>\n",
       "      <td>rep1_peak_496</td>\n",
       "      <td>33</td>\n",
       "      <td>.</td>\n",
       "      <td>2.96256</td>\n",
       "      <td>3.32397</td>\n",
       "      <td>0.942213</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL000216.2</td>\n",
       "      <td>29138</td>\n",
       "      <td>29276</td>\n",
       "      <td>rep1_peak_673</td>\n",
       "      <td>28</td>\n",
       "      <td>.</td>\n",
       "      <td>2.75165</td>\n",
       "      <td>2.81053</td>\n",
       "      <td>0.720430</td>\n",
       "      <td>118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL000224.1</td>\n",
       "      <td>31792</td>\n",
       "      <td>31992</td>\n",
       "      <td>rep1_peak_1045</td>\n",
       "      <td>13</td>\n",
       "      <td>.</td>\n",
       "      <td>1.75131</td>\n",
       "      <td>1.38178</td>\n",
       "      <td>0.251322</td>\n",
       "      <td>136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom  chromStart  chromEnd            name  score strand  \\\n",
       "0  GL000009.2      195221    195358   rep1_peak_216     28      .   \n",
       "1  GL000194.1      155811    155924   rep1_peak_280     14      .   \n",
       "2  GL000208.1       47204     47410   rep1_peak_496     33      .   \n",
       "3  GL000216.2       29138     29276   rep1_peak_673     28      .   \n",
       "4  GL000224.1       31792     31992  rep1_peak_1045     13      .   \n",
       "\n",
       "   signalValue   pValue    qValue  peak  \n",
       "0      2.69423  2.80547  0.720430    75  \n",
       "1      1.55919  1.47921  0.251322   116  \n",
       "2      2.96256  3.32397  0.942213   167  \n",
       "3      2.75165  2.81053  0.720430   118  \n",
       "4      1.75131  1.38178  0.251322   136  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in peaks\n",
    "peaks = pd.read_table(\"results/filtered_peaks.bed\", header=None)\n",
    "\n",
    "# set colnames for readability\n",
    "peaks.columns = [\"chrom\", \"chromStart\", \"chromEnd\", \"name\", \"score\", \"strand\", \"signalValue\", \"pValue\", \"qValue\", \"peak\"]\n",
    "peaks.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>RAD9A</td>\n",
       "      <td>NM_001243224,NM_004584</td>\n",
       "      <td>-0.994436</td>\n",
       "      <td>4.514560e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TMEM70</td>\n",
       "      <td>NM_001040613,NM_017866,NR_033334</td>\n",
       "      <td>-0.655475</td>\n",
       "      <td>1.825129e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>FAM133B</td>\n",
       "      <td>NM_001040057,NM_001288584,NM_152789,NR_109929</td>\n",
       "      <td>-0.521628</td>\n",
       "      <td>3.368896e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>LRRC20</td>\n",
       "      <td>NM_001278211,NM_001278212,NM_001278213,NM_0012...</td>\n",
       "      <td>0.646662</td>\n",
       "      <td>1.476895e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SLC7A5</td>\n",
       "      <td>NM_003486</td>\n",
       "      <td>-0.785875</td>\n",
       "      <td>7.129022e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_name                                         transcript  \\\n",
       "7      RAD9A                             NM_001243224,NM_004584   \n",
       "8     TMEM70                   NM_001040613,NM_017866,NR_033334   \n",
       "12   FAM133B      NM_001040057,NM_001288584,NM_152789,NR_109929   \n",
       "14    LRRC20  NM_001278211,NM_001278212,NM_001278213,NM_0012...   \n",
       "22    SLC7A5                                          NM_003486   \n",
       "\n",
       "    log2FoldChange          padj  \n",
       "7        -0.994436  4.514560e-06  \n",
       "8        -0.655475  1.825129e-03  \n",
       "12       -0.521628  3.368896e-02  \n",
       "14        0.646662  1.476895e-02  \n",
       "22       -0.785875  7.129022e-09  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in rnaseq res\n",
    "rnaseq = pd.read_table(\"rnaseq_res.txt\")\n",
    "\n",
    "# filter by padj\n",
    "pcut = 0.05\n",
    "rnaseq_fltr = rnaseq[rnaseq['padj'] < 0.05]\n",
    "rnaseq_fltr.columns = ['gene_name', 'transcript', 'log2FoldChange', 'padj']\n",
    "rnaseq_fltr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/ipykernel_3613278/4222830413.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  genes['gene_name'] = genes['attribute'].str.extract('gene_name \"([^\"]+)\"')\n",
      "/scratch/ipykernel_3613278/4222830413.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  genes['gene_id'] = genes['attribute'].str.extract('gene_id \"([^\"]+)\"')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000290825.1\"; gene_type \"lncRNA...</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.6\"; gene_type \"transc...</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENSG00000223972.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14696</td>\n",
       "      <td>24886</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.6\"; gene_type \"unproc...</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENSG00000227232.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278267.1\"; gene_type \"miRNA\"...</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243485.5\"; gene_type \"lncRNA...</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   source feature  start    end score strand frame  \\\n",
       "0   chr1   HAVANA    gene  11869  14409     .      +     .   \n",
       "5   chr1   HAVANA    gene  12010  13670     .      +     .   \n",
       "13  chr1   HAVANA    gene  14696  24886     .      -     .   \n",
       "25  chr1  ENSEMBL    gene  17369  17436     .      -     .   \n",
       "28  chr1   HAVANA    gene  29554  31109     .      +     .   \n",
       "\n",
       "                                            attribute    gene_name  \\\n",
       "0   gene_id \"ENSG00000290825.1\"; gene_type \"lncRNA...      DDX11L2   \n",
       "5   gene_id \"ENSG00000223972.6\"; gene_type \"transc...      DDX11L1   \n",
       "13  gene_id \"ENSG00000227232.6\"; gene_type \"unproc...       WASH7P   \n",
       "25  gene_id \"ENSG00000278267.1\"; gene_type \"miRNA\"...    MIR6859-1   \n",
       "28  gene_id \"ENSG00000243485.5\"; gene_type \"lncRNA...  MIR1302-2HG   \n",
       "\n",
       "              gene_id  \n",
       "0   ENSG00000290825.1  \n",
       "5   ENSG00000223972.6  \n",
       "13  ENSG00000227232.6  \n",
       "25  ENSG00000278267.1  \n",
       "28  ENSG00000243485.5  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a database from gtf\n",
    "gtf = pd.read_csv('/projectnb/bf528/materials/project-2-chipseq/refs/gencode.v45.primary_assembly.annotation.gtf', sep=\"\\t\", comment=\"#\", header=None)\n",
    "gtf.columns = ['chrom', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "\n",
    "# Filter for 'gene' entries\n",
    "genes = gtf[gtf['feature'] == 'gene']\n",
    "\n",
    "# Extract gene_name and gene_id\n",
    "genes['gene_name'] = genes['attribute'].str.extract('gene_name \"([^\"]+)\"')\n",
    "genes['gene_id'] = genes['attribute'].str.extract('gene_id \"([^\"]+)\"')\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>padj</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RAD9A</td>\n",
       "      <td>NM_001243224,NM_004584</td>\n",
       "      <td>-0.994436</td>\n",
       "      <td>4.514560e-06</td>\n",
       "      <td>chr11</td>\n",
       "      <td>67317871</td>\n",
       "      <td>67398410</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TMEM70</td>\n",
       "      <td>NM_001040613,NM_017866,NR_033334</td>\n",
       "      <td>-0.655475</td>\n",
       "      <td>1.825129e-03</td>\n",
       "      <td>chr8</td>\n",
       "      <td>73972437</td>\n",
       "      <td>73982783</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FAM133B</td>\n",
       "      <td>NM_001040057,NM_001288584,NM_152789,NR_109929</td>\n",
       "      <td>-0.521628</td>\n",
       "      <td>3.368896e-02</td>\n",
       "      <td>chr7</td>\n",
       "      <td>92560758</td>\n",
       "      <td>92590393</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LRRC20</td>\n",
       "      <td>NM_001278211,NM_001278212,NM_001278213,NM_0012...</td>\n",
       "      <td>0.646662</td>\n",
       "      <td>1.476895e-02</td>\n",
       "      <td>chr10</td>\n",
       "      <td>70298970</td>\n",
       "      <td>70382650</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SLC7A5</td>\n",
       "      <td>NM_003486</td>\n",
       "      <td>-0.785875</td>\n",
       "      <td>7.129022e-09</td>\n",
       "      <td>chr16</td>\n",
       "      <td>87830023</td>\n",
       "      <td>87869507</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name                                         transcript  \\\n",
       "0     RAD9A                             NM_001243224,NM_004584   \n",
       "1    TMEM70                   NM_001040613,NM_017866,NR_033334   \n",
       "2   FAM133B      NM_001040057,NM_001288584,NM_152789,NR_109929   \n",
       "3    LRRC20  NM_001278211,NM_001278212,NM_001278213,NM_0012...   \n",
       "4    SLC7A5                                          NM_003486   \n",
       "\n",
       "   log2FoldChange          padj  chrom     start       end strand  \n",
       "0       -0.994436  4.514560e-06  chr11  67317871  67398410      +  \n",
       "1       -0.655475  1.825129e-03   chr8  73972437  73982783      +  \n",
       "2       -0.521628  3.368896e-02   chr7  92560758  92590393      -  \n",
       "3        0.646662  1.476895e-02  chr10  70298970  70382650      -  \n",
       "4       -0.785875  7.129022e-09  chr16  87830023  87869507      -  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge rnaseq and gtf\n",
    "merged = rnaseq_fltr.merge(genes[['chrom', 'start', 'end', 'strand', 'gene_name']], on='gene_name', how='inner')\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/ipykernel_3613278/3656577109.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_bed['start'] = gene_bed['start'].astype(int)\n",
      "/scratch/ipykernel_3613278/3656577109.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_bed['end'] = gene_bed['end'].astype(int)\n",
      "***** WARNING: File /scratch/pybedtools.0baylfq7.tmp has inconsistent naming convention for record:\n",
      "chr1\t1649716\t1649856\trep1_peak_7020\n",
      "\n",
      "***** WARNING: File /scratch/pybedtools.0baylfq7.tmp has inconsistent naming convention for record:\n",
      "chr1\t1649716\t1649856\trep1_peak_7020\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_chrom</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>peak_chrom</th>\n",
       "      <th>peak_start</th>\n",
       "      <th>peak_end</th>\n",
       "      <th>peak_name</th>\n",
       "      <th>transcript</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>67317871</td>\n",
       "      <td>67398410</td>\n",
       "      <td>RAD9A</td>\n",
       "      <td>chr11</td>\n",
       "      <td>67335729</td>\n",
       "      <td>67335891</td>\n",
       "      <td>rep1_peak_187956</td>\n",
       "      <td>NM_001243224,NM_004584</td>\n",
       "      <td>-0.994436</td>\n",
       "      <td>0.000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>67317871</td>\n",
       "      <td>67398410</td>\n",
       "      <td>RAD9A</td>\n",
       "      <td>chr11</td>\n",
       "      <td>67383261</td>\n",
       "      <td>67383505</td>\n",
       "      <td>rep1_peak_187988</td>\n",
       "      <td>NM_001243224,NM_004584</td>\n",
       "      <td>-0.994436</td>\n",
       "      <td>0.000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>121989163</td>\n",
       "      <td>122254545</td>\n",
       "      <td>TACC2</td>\n",
       "      <td>chr10</td>\n",
       "      <td>122158200</td>\n",
       "      <td>122158410</td>\n",
       "      <td>rep1_peak_154761</td>\n",
       "      <td>NM_001291876,NM_001291877,NM_001291878,NM_0012...</td>\n",
       "      <td>-0.780570</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr3</td>\n",
       "      <td>66378797</td>\n",
       "      <td>66501263</td>\n",
       "      <td>LRIG1</td>\n",
       "      <td>chr3</td>\n",
       "      <td>66465422</td>\n",
       "      <td>66465558</td>\n",
       "      <td>rep1_peak_725040</td>\n",
       "      <td>NM_015541</td>\n",
       "      <td>-0.441669</td>\n",
       "      <td>0.008332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15940550</td>\n",
       "      <td>15947007</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>15945038</td>\n",
       "      <td>15945162</td>\n",
       "      <td>rep1_peak_536093</td>\n",
       "      <td>NM_001293228,NM_001293231,NM_001293233</td>\n",
       "      <td>1.961676</td>\n",
       "      <td>0.000093</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_chrom  gene_start   gene_end gene_name peak_chrom  peak_start  \\\n",
       "0      chr11    67317871   67398410     RAD9A      chr11    67335729   \n",
       "1      chr11    67317871   67398410     RAD9A      chr11    67383261   \n",
       "2      chr10   121989163  122254545     TACC2      chr10   122158200   \n",
       "3       chr3    66378797   66501263     LRIG1       chr3    66465422   \n",
       "4       chr2    15940550   15947007      MYCN       chr2    15945038   \n",
       "\n",
       "    peak_end         peak_name  \\\n",
       "0   67335891  rep1_peak_187956   \n",
       "1   67383505  rep1_peak_187988   \n",
       "2  122158410  rep1_peak_154761   \n",
       "3   66465558  rep1_peak_725040   \n",
       "4   15945162  rep1_peak_536093   \n",
       "\n",
       "                                          transcript  log2FoldChange      padj  \n",
       "0                             NM_001243224,NM_004584       -0.994436  0.000005  \n",
       "1                             NM_001243224,NM_004584       -0.994436  0.000005  \n",
       "2  NM_001291876,NM_001291877,NM_001291878,NM_0012...       -0.780570  0.000004  \n",
       "3                                          NM_015541       -0.441669  0.008332  \n",
       "4             NM_001293228,NM_001293231,NM_001293233        1.961676  0.000093  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_bed = merged[['chrom', 'start', 'end', 'gene_name']]\n",
    "gene_bed['start'] = gene_bed['start'].astype(int)\n",
    "gene_bed['end'] = gene_bed['end'].astype(int)\n",
    "\n",
    "peak_bed = peaks[['chrom', 'chromStart', 'chromEnd', 'name']]\n",
    "peak_bed.columns = ['chrom', 'start', 'end', 'name']\n",
    "\n",
    "gene_bt = pybedtools.BedTool.from_dataframe(gene_bed)\n",
    "peak_bt = pybedtools.BedTool.from_dataframe(peak_bed)\n",
    "\n",
    "overlaps = gene_bt.intersect(peak_bt, wa=True, wb=True)\n",
    "overlap_df = overlaps.to_dataframe(names=[\n",
    "    'gene_chrom', 'gene_start', 'gene_end', 'gene_name',\n",
    "    'peak_chrom', 'peak_start', 'peak_end', 'peak_name'\n",
    "])\n",
    "final_df = overlap_df.merge(rnaseq_fltr, on='gene_name', how='left')\n",
    "final_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "proj2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
